### DISCUSSION TITLE
Producer(auto_declare=True).publish() semantics

### DISCUSSION THREAD

**Discussion by povilasb** [celery/kombu#1530] [Ideas]
## The problem

Currently `Producer` with `auto_declare` set may have some unexpected semantics, at least to me :)

Here's a very basic example that most people would have a good sense of what's going to happen:
```py
Producer(auto_declare=True).publish(b"message")
```
It will make sure the default exchange is declared before sending the message to.

However, consider such code:
```py
Producer(auto_declare=True).publish(b"message", exchange="tasks")
```

If not reading the `.publish()` docs one could misinterpret the outcome
```py
            exchange (kombu.entity.Exchange, str): Override the exchange.
                Note that this exchange must have been declared.
```

So now `Producer` [will still try to declare the default exchange](https://github.com/celery/kombu/blob/a93099b671f83ed5a3f723198cb821041ca5e13e/kombu/messaging.py#L174) but publish to the "tasks" exchange:
```py
        if self.auto_declare and self.exchange.name:
            if self.exchange not in declare:
                # XXX declare should be a Set.
                declare.append(self.exchange)
```

## Potential change

What if `auto_declare` would try to declare the specified exchange instead?
1. default one if `.publish()` has no exchange specified
2. the one specified by `.publish()` otherwise

The same way the exchange name is already resolved before publishing:
```py
        exchange_name, properties['delivery_mode'] = self._delivery_details(
            exchange or self.exchange, delivery_mode,
        )
```

**Comment by TayfurYldz** [comment]
Your reading still matches the current implementation. `Producer.publish()` first handles `auto_declare` using `self.exchange`: if `self.exchange.name` is non-empty and it isn't already in `declare`, that default exchange is appended. The per-call `exchange` argument is resolved later for the actual publish operation; it does not replace the entity selected by that auto-declare block.

So `Producer(default_exchange, auto_declare=True).publish(..., exchange="tasks")` should not be read as "declare `tasks` automatically". In particular, a string exchange override has no declaration metadata such as exchange type/durability anyway.

If the caller needs declaration, the unambiguous current API is to create the `Exchange` entity and pass it through `declare=[exchange]` (or declare it explicitly) before publishing. Changing `auto_declare` to follow a per-call override would need to define behavior separately for an `Exchange` object versus a bare string and would be a semantic/API change, not just a bug fix.
